### PR TITLE
rotate module

### LINKED ISSUE #1137: Rotate module not functioning correctly
The rotate module implemented currently is not working correctly. I don't think this is some bug but the way it has been implemented. Right now, after rotating the image, it acquires the same dimensions as the original image due to which sometimes a rectangular image gets cropped which should not be the case.

CC: @jywarren @tech4GT @HarshKhandeparkar @aashna27

### PR DESCRIPTION
**Pull request by keshav234156**
Fixes #1137 
 this PR fixes the  clipping of image while using rotate module
Screenshot from 2019-08-20 13-21-01

image at 30
download (17)

image at 90
download (18)

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

* [ ] tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR -- or run tests locally with `npm test`
* [ ] code is in uniquely-named feature branch and has no merge conflicts
* [ ] PR is descriptively titled
* [x] ask `@publiclab/is-reviewers` for help, in a comment below

> We're happy to help you get this ready -- don't be afraid to ask for help, and **don't be discouraged** if your tests fail at first!

If tests do fail, click on the red `X` to learn why by reading the logs.

Please be sure you've reviewed our contribution guidelines at https://publiclab.org/contributing-to-public-lab-software

Thanks!

### CODE REVIEW

**Comment by keshav234156** [comment]
@jywarren @HarshKhandeparkar  can you please guide me how to fix the issues given by codeclimate

**Comment by HarshKhandeparkar** [comment]
I never fixed them till date. They will require a separate dedicated PR.
don't worry about it, they can be ignored for now.

**Comment by aashna27** [comment]
Yeah just make sure that the new code you add is refactored.

**Comment by keshav234156** [comment]
@HarshKhandeparkar @jywarren can you please review the changes!!

**Comment by HarshKhandeparkar** [comment]
Nice work! I am a bit busy as of now. I guess I will review it tmrw. I'll
have to test it out locally. Slowly but surely.

**Comment by keshav234156** [comment]
@HarshKhandeparkar can you please review if you are free

**Comment by HarshKhandeparkar** [comment]
I am extremely sorry. I have a test on this Wednesday. I will only be able
to review on that day. I will do it as soon as possible.

**Comment by keshav234156** [comment]
@HarshKhandeparkar Splitting the file inti Module.js and Rotate.js would make the code  easy to understand and i think there won't be any issue.So will  make change and push it asap.

**Comment by keshav234156** [comment]
@jywarren @HarshKhandeparkar have splitted it into 2 files.please have a look

File: src/modules/Rotate/Module.js [hunk 1]
```diff
@@ -23,20 +23,17 @@ module.exports = function Rotate(options, UI) {
...
-
-      var rotated = bitmap.rotate({
-        degrees: rotate_value,
-      });
-      pixels.data = rotated._data.data;
-
+      var radians = 3.141592653589793 * rotate_value / 180;
+      var width = pixels.shape[0];
+      var height = pixels.shape[1];
+      var cos = Math.cos(radians);
```
Comment (HarshKhandeparkar) [inline src/modules/Rotate/Module.js]: If you have split this into two files, is this still required?

File: src/modules/Rotate/Rotate.js [hunk 2]
```diff
@@ -0,0 +1,54 @@
+module.exports = function Rotate(pixels, pixels2, options){
+  var imagejs = require('imagejs');
+  var rotate_value = (options.rotate) % 360;
+  var radians = 3.141592653589793 * rotate_value / 180;
```
Comment (HarshKhandeparkar) [inline src/modules/Rotate/Rotate.js]: You can use `Math.PI` constant

File: src/modules/Rotate/Module.js (hunk 1 above)
Comment (keshav234156) [reply src/modules/Rotate/Module.js]: @HarshKhandeparkar i think these would still be required as we are returning pixels2 from the Rotate.js ,so it must be initialised before passing it as argument in module.js.
What do you think about this?

File: src/modules/Rotate/Rotate.js (hunk 2 above)
Comment (keshav234156) [reply src/modules/Rotate/Rotate.js]: Ok i think it good to use math.pi

**Comment by HarshKhandeparkar** [comment]
Then why is it done again in Rotate.js?

**Comment by keshav234156** [comment]
@publiclab/reviewers @jywarren can you please review

**Comment by aashna27** [comment]
The work done is really great. Thanks. Can you document the flow so that it becomes easy to understand.

**Comment by keshav234156** [comment]
@aashna27 have a look..

**Comment by keshav234156** [comment]
@jywarren  Have a look

**Comment by keshav234156** [comment]
@aashna27 @jywarren can you please review this PR!!

**Comment by jywarren** [comment]
Apologies, @keshav234156 i was in a week of meetings, I'm looking now! First, i'll update to sync with the latest main branch. Then will look through the code. Thanks for your patience, we're still recovering a bit from our Summer of Code wrap-up!

File: src/modules/Rotate/Rotate.js [hunk 3]
```diff
@@ -0,0 +1,46 @@
+module.exports = function Rotate(pixels, pixels2, options, rotate_value, width, height, cos, sin){
+  var imagejs = require('imagejs');
+  var height_half = Math.floor(height / 2);
+  var width_half = Math.floor(width / 2);
+  var dimension = width + height;
+
+  if (rotate_value % 360 == 0)
+    return pixels;
+  function copyPixel(x1, y1, x2, y2){
```
Comment (jywarren) [inline src/modules/Rotate/Rotate.js]: Hi, couldn't this get extra `pixels1, pixels2` parameters and then we'd have only one function instead of 2?

File: src/modules/Rotate/Rotate.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+  for (var n = 0; n < pixels.shape[0]; n++){
+    for (var m = 0; m < pixels.shape[1]; m++){
+      copyPixel(n + height_half, m + width_half, n, m);
+    }
+  }
+  //rotating pixels1
+  var bitmap = new imagejs.Bitmap({ width: pixels1.shape[0], height: pixels1.shape[1] });
+  bitmap._data.data = pixels1.data;
+
+  var rotated = bitmap.rotate({
```
Comment (jywarren) [inline src/modules/Rotate/Rotate.js]: Hmm, so why do we need to do so much pixel copying and rotation math if we are still using the Imagejs Bitmap function to perform the rotation? Perhaps some comments would help clarify what's going on here? Sorry, it's just a little hard to follow! Thanks for your help, i love the output, just being sure I understand what's happening inside as well :-)

Comment (keshav234156) [reply src/modules/Rotate/Rotate.js]: earlier if we use bitmap on original image then it cropped some of the image.to solve this i have put the image on a larger frame ie is square  so that there is no problem of clipping and then extra white is cropped according to to final image dimensions after rotation

Comment (keshav234156) [reply src/modules/Rotate/Rotate.js]: pixel1 is a square frame with original image at it centre,then pixel1 is rotated with the help of imagejs.at last we have have copied pixels from pixels1 to pixel2 so that there is no whitespace.

Comment (keshav234156) [reply src/modules/Rotate/Rotate.js]: pixel1 is a square frame that contain original image at it's centre.the we rotated the image with help of imagejs,at the last we copied pixels from pixel1 to pixel2 so that there is no extra whitespace.

File: src/modules/Rotate/Rotate.js (hunk 3 above)
Comment (keshav234156) [reply src/modules/Rotate/Rotate.js]: ok, would do the changes to make a 1 function.

Comment (keshav234156) [reply src/modules/Rotate/Rotate.js]: @jywarren any other changes that are required??

Comment (keshav234156) [reply src/modules/Rotate/Rotate.js]: @jywarren Had done the changes to make a single function instead of 2

**Comment by keshav234156** [comment]
@jywarren if there any other changes or method in your mind please let me know!!.would love to solve this issue.

**Comment by jywarren** [comment]
Looks great. Thanks so much!
